src/app/clusters/joint-fabric-datastore-server/joint-fabric-datastore-server.cpp

```diff
@@ -127,43 +182,142 @@ bool emberAfJointFabricDatastoreClusterAddGroupCallback(
     return true;
 }
 
-// TODO
 bool emberAfJointFabricDatastoreClusterAddKeySetCallback(
     CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
     const JointFabricDatastoreCluster::Commands::AddKeySet::DecodableType & commandData)
 {
+    CHIP_ERROR err                                                                              = CHIP_NO_ERROR;
+    JointFabricDatastoreCluster::Structs::DatastoreGroupKeySetStruct::DecodableType groupKeySet = commandData.groupKeySet;
+    JointFabricDatastorage & JointFabricDatastorage = Server::GetInstance().GetJointFabricDatastorage();
+
+    VerifyOrExit(JointFabricDatastorage.IsGroupKeySetEntryPresent(groupKeySet.groupKeySetID) == false,
+                 err = CHIP_ERROR_INVALID_ARGUMENT);
+    SuccessOrExit(err = JointFabricDatastorage.AddGroupKeySetEntry(groupKeySet));
+
+exit:
+    if (err == CHIP_NO_ERROR)
+    {
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Success);
+    }
+    else
+    {
+        ChipLogError(DataManagement, "JointFabricDatastoreCluster: failed with error: %" CHIP_ERROR_FORMAT, err.Format());
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::ClusterStatusCode(err));
+    }
+
     return true;
 }
 
-// TODO
 bool emberAfJointFabricDatastoreClusterRemoveNodeCallback(
     CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
     const JointFabricDatastoreCluster::Commands::RemoveNode::DecodableType & commandData)
 {
+    NodeId nodeId = commandData.nodeID;
+
+    if (commandPath.mEndpointId != kRootEndpointId)
+    {
+        ChipLogError(DataManagement, "JointFabricDatastoreCluster: invalid endpoint in RemoveNode request");
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::InvalidCommand);
+        return true;
+    }
+
+    CHIP_ERROR err = Server::GetInstance().GetJointFabricDatastorage().RemoveNode(nodeId);
+
+    if (err == CHIP_NO_ERROR)
+    {
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Success);
+    }
+    else
+    {
+        ChipLogError(DataManagement, "JointFabricDatastoreCluster: failed with error: %" CHIP_ERROR_FORMAT, err.Format());
+        commandObj->AddStatus(commandPath, Protocols::InteractionModel::ClusterStatusCode(err));
+    }
+
     return true;
 }
 
-// TODO
 bool emberAfJointFabricDatastoreClusterUpdateNodeCallback(
     CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
     const JointFabricDatastoreCluster::Commands::UpdateNode::DecodableType & commandData)
 {
+    NodeId nodeId                 = commandData.nodeID;
+    const CharSpan & friendlyName = commandData.friendlyName;
+
+    if (commandPath.mEndpointId != kRootEndpointId)
```

Comment:
same here (JF Admin endpoint is probably not the root endpoint)

Reply:
I won't repeat this for all of the commands...

Reply:
Thanks. All instances of this endpoint check have been removed in the latest commit.